File: 剑指 Offer（第 2 版）/18.cpp
// 剑指 Offer 18. 删除链表的节点
// https://leetcode-cn.com/problems/shan-chu-lian-biao-de-jie-dian-lcof/


using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
    ListNode* deleteNode(ListNode* head, int val) {
        if(head->val==val){
            return head->next;
        }
        ListNode* pre = head;
        ListNode* cur = head->next;
        while(cur){
            if(cur->val==val){
                pre->next = cur->next;
                break;
            }
            pre = cur;
            cur = cur->next;
        }
        return head;
    }
};

File: 剑指 Offer（第 2 版）/18.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

class Solution {
    public ListNode deleteNode(ListNode head, int val) {
        if(head.val==val) return head.next;
        ListNode cur = head.next;
        ListNode prev = head;
        while(cur!=null && cur.val!=val){
            cur = cur.next;
            prev = prev.next;
        }
        prev.next = cur.next;
        return head;
    }
}